core/pta/widevine.c

```diff
@@ -0,0 +1,142 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2023, The ChromiumOS Authors
+ */
+
+#include <compiler.h>
+#include <initcall.h>
+#include <kernel/boot.h>
+#include <kernel/dt.h>
+#include <kernel/pseudo_ta.h>
+#include <libfdt.h>
+#include <malloc.h>
+#include <mm/tee_mm.h>
+#include <mm/tee_pager.h>
+#include <stdio.h>
+#include <string.h>
+#include <string_ext.h>
+#include <tee_api.h>
+#include <trace.h>
+
+#define PTA_NAME "widevine.pta"
+
+#define GET_TPM_PUBKEY 0
+#define GET_WIDEVINE_PRIVKEY 1
+
+/* 721f4da9-da05-40d4-a1a3-8377c1e08b0a */
+#define WIDEVINE_PTA_UUID                                              \
+	{                                                              \
+		0x721f4da9, 0xda05, 0x40d4,                            \
+		{                                                      \
+			0xa1, 0xa3, 0x83, 0x77, 0xc1, 0xe0, 0x8b, 0x0a \
+		}                                                      \
+	}
+
+#define TPM_AUTH_PUB_MAX_SIZE 512
+#define WIDEVINE_PRIV_MAX_SIZE 32
+
+/* The public key to encrypt the communication to TPM. */
+static uint8_t tpm_auth_pub[TPM_AUTH_PUB_MAX_SIZE];
+static uint32_t tpm_auth_pub_size;
+
+/* The private key for Widevine signing. */
+static uint8_t widevine_priv[WIDEVINE_PRIV_MAX_SIZE];
+static uint32_t widevine_priv_size;
+
+static TEE_Result init_widevine_dt_data(void)
+{
+	int node = 0;
+	int len = 0;
+	void *fdt;
+	const void *result = NULL;
+
+	fdt = get_dt();
+
+	if (!fdt)
+		return TEE_ERROR_NO_DATA;
+
+	node = fdt_path_offset(fdt, "/options/widevine");
+	if (node < 0)
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	result = fdt_getprop(fdt, node, "tcg-tpm-auth-public-key", &len);
+	if (!result)
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	if (len > TPM_AUTH_PUB_MAX_SIZE)
+		return TEE_ERROR_OVERFLOW;
+
+	memcpy(tpm_auth_pub, result, len);
+	tpm_auth_pub_size = len;
+
+	result = fdt_getprop(fdt, node, "widevine-root-of-trust", &len);
+	if (!result)
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	if (len > WIDEVINE_PRIV_MAX_SIZE)
+		return TEE_ERROR_OVERFLOW;
+
+	memcpy(widevine_priv, result, len);
+	widevine_priv_size = len;
+
+	return TEE_SUCCESS;
+}
+
+service_init(init_widevine_dt_data);
+
+static TEE_Result get_dt_data(uint32_t ptypes, TEE_Param params[TEE_NUM_PARAMS],
+			      uint32_t cmd)
+{
+	uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INOUT,
+					  TEE_PARAM_TYPE_VALUE_OUTPUT,
+					  TEE_PARAM_TYPE_NONE,
+					  TEE_PARAM_TYPE_NONE);
+
+	uint8_t *data;
+	uint32_t data_length;
+
+	if (exp_pt != ptypes)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	/* Make sure we called from Secure World */
+	if (!ts_get_calling_session())
+		return TEE_ERROR_ACCESS_DENIED;
+
+	if (cmd == GET_TPM_PUBKEY) {
+		data = tpm_auth_pub;
+		data_length = tpm_auth_pub_size;
+	} else if (cmd == GET_WIDEVINE_PRIVKEY) {
+		data = widevine_priv;
+		data_length = widevine_priv_size;
```

Comment:
This works for a system where we can guarantee that no other TA's are deployed. If other TA's could be deployed, then another TA could simply call this TA and just get the private key. I don't think we have something to support this, but it'd great if we could control and verify who is calling this function. @jenswi-linaro , any ideas?

Reply:
There is at least a way to check the calling TA has the expected UUID, something like:
```c
	if (to_ta_session(ts_get_current_session())->clnt_id.login != TEE_LOGIN_TRUSTED_APP ||
	    memcmp(&ts_get_calling_session()->ctx->uuid, (char *)&expected_ta_uuid, sizeof(TEE_UUID))) {
		return TEE_ERROR_ACCESS_DENIED;
	}
```
Using this in conjunction with TA namespaces (see doc on [verifying-with-subkeys](https://optee.readthedocs.io/en/latest/architecture/trusted_applications.html#verifying-with-subkeys)) allows specific PTAs to be reachable only from specific TAs.